src/OrleansRuntime/Streams/QueueBalancer/LeaseBasedQueueBalancer.cs

```diff
@@ -0,0 +1,345 @@
+﻿using Orleans.LeaseProviders;
+using Orleans.Runtime;
+using Orleans.Runtime.Configuration;
+using Orleans.Streams;
+using System;
+using System.Collections.Concurrent;
+using System.Collections.Generic;
+using System.Collections.ObjectModel;
+using System.Linq;
+using System.Text;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.Extensions.DependencyInjection;
+
+namespace Orleans.Streams
+{
+    /// <summary>
+    /// IResourceSelector selects a centain amount of resource from a resource list
+    /// </summary>
+    /// <typeparam name="T"></typeparam>
+    internal interface IResourceSelector<T>
+    {
+        /// <summary>
+        /// Try to select certain count of resources from resource list, which doesn't overlap with existing selection
+        /// </summary>
+        /// <param name="newSelectionCount"></param>
+        /// <param name="existingSelection"></param>
+        /// <returns></returns>
+        IEnumerable<T> NextSelection(int newSelectionCount, IEnumerable<T> existingSelection);
+    }
+
+    /// <summary>
+    /// Selector using round robin algorithm
+    /// </summary>
+    /// <typeparam name="T"></typeparam>
+    internal class RoundRobinSelector<T> : IResourceSelector<T>
+    {
+        private List<T> resources;
+        private int lastSelection;
+        public RoundRobinSelector(List<T> resources)
+        {
+            this.resources = resources;
+            this.lastSelection = -1;
+        }
+
+        /// <summary>
+        /// Try to select certain count of resources from resource list, which doesn't overlap with existing resources
+        /// </summary>
+        /// <param name="newSelectionCount"></param>
+        /// <param name="existingSelection"></param>
+        /// <returns></returns>
+        public IEnumerable<T> NextSelection(int newSelectionCount, IEnumerable<T> existingSelection)
```

Comment:
The constructor takes a `List<T>`, but this method takes/returns an `IEnumerable<T>` and then enumerates it many times (allocating an enumerator every time). Suggest you use `List<T>` instead of `IEnumerable<T>`. Alternatively, given this method works on contiguous collections, you only need to find the first and last item (first + count) in the existing selection and avoid all items in that range. So you could have a 'SelectionSpan` type which recirds that info.